lib/index.js

```diff
@@ -0,0 +1,211 @@
+
+  // the genesis node functions as a dictator until the electorPool document
+  // is established
+  if(electorPoolElectors.length === 0) {
+    return {electors: [{id: consensusProofPeers[0]}], recoveryElectors: []};
+  }
+
+  const {blockHash} = latestBlockSummary.eventBlock.meta;
+  const baseHashBuffer = Buffer.from(blockHash);
+  // the hash of the previous block is combined with the elector id to
+  // prevent any elector from *always* being sorted to the top
+  electorPoolElectors.sort((a, b) => {
+    // generate and cache hashes
+    a._hashBuffer = a._hashBuffer || xor(baseHashBuffer, Buffer.from(a.id));
+    b._hashBuffer = b._hashBuffer || xor(baseHashBuffer, Buffer.from(b.id));
+
+    // sort by hash
+    return Buffer.compare(a._hashBuffer, b._hashBuffer);
+  });
+
+  // TODO: allow the electorPool document to specify an electorCount. Use the
+  // value from the electorPool document first if available and if absent,
+  // use the value specified in the ledgerConfiguration.
+
+  let electorCount = Math.min(
+    ledgerConfiguration.electorSelectionMethod.electorCount,
+    electorPoolElectors.length);
+
+  // adjust electorCount to the form 3f+1
+  electorCount = _computeTargetElectorCount(
+    {coefficient: 3, originalCount: electorCount});
+
+  const recoveryElectorCount = _computeTargetElectorCount(
+    {coefficient: 1, originalCount: electorCount});
+
+  const electors = [];
+  const recoveryElectors = [];
+  const recoveryElectorCandidates = [];
+  const guarantorElectorCandidates = [];
+  const regularElectorCandidates = [];
+
+  // sort electorPoolElectors into categories based on elector.type
+  let i = electorPoolElectors.length;
+  while(i--) {
+    const elector = electorPoolElectors.pop();
+    const electorType = [].concat(elector.type);
+    if(electorType.includes(ELECTOR_TYPE_RECOVERY)) {
+      recoveryElectorCandidates.push(elector);
+    } else if(electorType.includes(ELECTOR_TYPE_GUARANTOR)) {
+      guarantorElectorCandidates.push(elector);
+    } else if(electorType.includes(ELECTOR_TYPE_REGULAR)) {
+      regularElectorCandidates.push(elector);
+    }
+  }
+
+  // populate electors and recoveryElectors
+  // populate electors by:
+  //   1. use all recoveryElectorCandidates
+  //   2. use all guarantorElectorCandidates
+  //   3. use regularElectorCandidates to fill remaining spots
+  while(electors.length < electorCount) {
+    let elector;
+    let recoveryElectorCandidate = false;
+    if(recoveryElectorCandidates.length > 0) {
+      elector = recoveryElectorCandidates.pop();
+      recoveryElectorCandidate = true;
+    } else if(guarantorElectorCandidates.length > 0) {
+      elector = guarantorElectorCandidates.pop();
+    } else if(regularElectorCandidates.length > 0) {
+      // TODO: possibly prefer cadidates in this class that are participating?
+      elector = regularElectorCandidates.pop();
+    }
+    electors.push({id: elector.id});
```

Comment:
This seems like it would fill up the electors with more recovery electors than is necessary? We only need up to `f+1` recovery electors, after that we want to pick regular ones at pseudorandom from those that have been participating recently.

Reply:
Pseudorandom sorting can mix in previous block hash (i.e. use the code from the other selection mechanisms).

Reply:
Again, all electors have already been sorted in this fashion,  So it should just be a matter of pulling out of the proper bucket.